File: internships/tinkoff/25_spring/F.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int main() {
    std::ios_base::sync_with_stdio(0);
    std::cin.tie(0);
    std::cout.tie(0);

    int n;
    cin >> n;
    vector<pair<int, int>> dots(n);

    for (int i = 0; i < n; ++i) {
        cin >> dots[i].first >> dots[i].second;
    }

    int max_collinear = 0;
    for (int i = 0; i < n; ++i) {
        for (int j = i + 1; j < n; ++j) {
            int collinear = 0;
            for (int k = 0; k < n; ++k) {
                if ((dots[k].first - dots[i].first) * (dots[j].second - dots[i].second) ==
                    (dots[k].second - dots[i].second) * (dots[j].first - dots[i].first)) {
                    collinear++;
                }
            }
            max_collinear = max(max_collinear, collinear);
        }
    }

    if (max_collinear <= (2 * n) / 3) {
        cout << n / 3 << endl;
    } else {
        cout << n - max_collinear << endl;
    }

    return 0;
}

File: internships/tinkoff/25_spring/F.py
import sys


def main() -> None:
    n = int(sys.stdin.readline().rstrip())
    dots = [tuple(map(int, sys.stdin.readline().split())) for _ in range(n)]

    max_collinear = 0
    for i in range(n):
        for j in range(i + 1, n):
            collinear = 0
            for k in range(n):
                if (dots[k][0] - dots[i][0]) * (dots[j][1] - dots[i][1]) == (dots[k][1] - dots[i][1]) * (dots[j][0] - dots[i][0]):
                    collinear += 1
            max_collinear = max(max_collinear, collinear)

    if max_collinear <= (2 * n) // 3:
        print(n // 3)
    else:
        print(n - max_collinear)


if __name__ == "__main__":
    main()
